File: frontend/lib/api/quantumAPI.ts
/**
 * Quantum Command Center - Frontend API Integration
 * Connects frontend components to backend API endpoints
 */

import { apiService, ApiResponse } from '../api/apiService';

export interface QuantumSystem {
  name: string;
  status: string;
  qubits: number;
  queue_length: number;
  avg_wait_time: string;
}

export interface HPCCluster {
  name: string;
  status: string;
  nodes: number;
  cpu_utilization: number;
  memory_utilization: number;
}

export interface QuantumJob {
  job_id: string;
  status: string;
  progress: number;
  quantum_qubits: number;
  hpc_utilization: number;
  error_message?: string;
}

export interface SystemMetrics {
  quantum_metrics: {
    total_jobs: number;
    success_rate: number;
    avg_execution_time: string;
    queue_wait_time: string;
  };
  hpc_metrics: {
    total_jobs: number;
    success_rate: number;
    avg_execution_time: string;
    resource_utilization: number;
  };
  system_metrics: {
    api_response_time: string;
    database_connections: number;
    memory_usage: string;
    cpu_usage: string;
  };
}

export interface RemediationStatus {
  self_healing: {
    enabled: boolean;
    last_check: string;
    issues_detected: number;
    issues_resolved: number;
    success_rate: number;
  };
  remediation_actions: Array<{
    action_id: string;
    type: string;
    status: string;
    timestamp: string;
  }>;
  recommendations: string[];
}

class QuantumCommandCenterAPI {
  // Quantum Systems Management
  async getQuantumSystems(): Promise<ApiResponse<{ quantum_systems: Record<string, QuantumSystem> }>> {
    return apiService.getQuantumData();
  }

  async getHPCClusters(): Promise<ApiResponse<{ hpc_clusters: Record<string, HPCCluster> }>> {
    return apiService.request('/api/hpc-data');
  }

  // Job Management
  async submitQuantumJob(data: {
    quantum_provider: string;
    hpc_cluster: string;
    job_type: string;
    parameters: Record<string, any>;
  }): Promise<ApiResponse<QuantumJob>> {
    return apiService.request('/api/submit-job', {
      method: 'POST',
      body: JSON.stringify(data),
    });
  }

  async getJobStatus(jobId: string): Promise<ApiResponse<QuantumJob>> {
    return apiService.request(`/api/job-status/${jobId}`);
  }

  // System Health & Monitoring
  async getSystemHealth(): Promise<ApiResponse<{
    quantum_systems: Record<string, string>;
    hpc_clusters: Record<string, string>;
    api_status: string;
    last_updated: string;
  }>> {
    return apiService.request('/api/system-health');
  }

  async getMonitoringMetrics(): Promise<ApiResponse<SystemMetrics>> {
    return apiService.request('/api/monitoring/metrics');
  }

  async getRemediationStatus(): Promise<ApiResponse<RemediationStatus>> {
    return apiService.request('/api/remediation/status');
  }

  // Real-time Updates (WebSocket simulation)
  async startRealTimeUpdates(callback: (data: any) => void): Promise<void> {
    // Simulate real-time updates with polling
    const pollInterval = setInterval(async () => {
      try {
        const [health, metrics, remediation] = await Promise.all([
          this.getSystemHealth(),
          this.getMonitoringMetrics(),
          this.getRemediationStatus(),
        ]);

        callback({
          health: health.data,
          metrics: metrics.data,
          remediation: remediation.data,
          timestamp: new Date().toISOString(),
        });
      } catch (error) {
        console.error('Real-time update error:', error);
      }
    }, 5000); // Update every 5 seconds

    // Return cleanup function
    return () => clearInterval(pollInterval);
  }

  // Utility Methods
  async testConnection(): Promise<boolean> {
    try {
      const response = await apiService.healthCheck();
      return response.status === 200;
    } catch {
      return false;
    }
  }

  async verifyAuthentication(): Promise<boolean> {
    try {
      const response = await apiService.verifyApiKey();
      return response.data?.valid === true;
    } catch {
      return false;
    }
  }
}

// Export singleton instance
export const quantumAPI = new QuantumCommandCenterAPI();

// Export for React hooks
export const useQuantumAPI = () => {
  return {
    quantumAPI,
    testConnection: () => quantumAPI.testConnection(),
    verifyAuth: () => quantumAPI.verifyAuthentication(),
  };
};

File: backend/routers/quantum_api.py
"""
Quantum Brain API Endpoints
Core API endpoints for Quantum Command Center
"""

from fastapi import APIRouter, HTTPException, Depends, status
from pydantic import BaseModel
from typing import List, Optional, Dict, Any
from datetime import datetime
import asyncio
import json

# Create main API router
router = APIRouter(prefix="/api", tags=["quantum-api"])

# Pydantic models for request/response
class QuantumDataRequest(BaseModel):
    quantum_provider: str
    hpc_cluster: str
    job_type: str
    parameters: Dict[str, Any]

class QuantumDataResponse(BaseModel):
    job_id: str
    status: str
    quantum_data: Dict[str, Any]
    hpc_metrics: Dict[str, Any]
    timestamp: datetime
    estimated_completion: Optional[datetime] = None

class JobStatusResponse(BaseModel):
    job_id: str
    status: str
    progress: float
    quantum_qubits: int
    hpc_utilization: float
    error_message: Optional[str] = None

class SystemHealthResponse(BaseModel):
    quantum_systems: Dict[str, str]
    hpc_clusters: Dict[str, str]
    api_status: str
    last_updated: datetime

# Mock data for demonstration
mock_quantum_data = {
    "ibm_quantum": {
        "status": "online",
        "qubits": 127,
        "queue_length": 3,
        "avg_wait_time": "2.5 minutes"
    },
    "aws_braket": {
        "status": "online", 
        "qubits": 32,
        "queue_length": 1,
        "avg_wait_time": "1.2 minutes"
    }
}

mock_hpc_data = {
    "aws_hpc": {
        "status": "online",
        "nodes": 100,
        "cpu_utilization": 75.5,
        "memory_utilization": 68.2
    },
    "azure_hpc": {
        "status": "online",
        "nodes": 50,
        "cpu_utilization": 45.8,
        "memory_utilization": 52.1
    }
}

# API Endpoints

@router.get("/quantum-data", response_model=Dict[str, Any])
async def get_quantum_data():
    """
    Get current quantum computing system status and data
    """
    return {
        "quantum_systems": mock_quantum_data,
        "timestamp": datetime.now(),
        "total_available_qubits": sum(system["qubits"] for system in mock_quantum_data.values()),
        "active_jobs": 4
    }

@router.get("/hpc-data", response_model=Dict[str, Any])
async def get_hpc_data():
    """
    Get current HPC cluster status and metrics
    """
    return {
        "hpc_clusters": mock_hpc_data,
        "timestamp": datetime.now(),
        "total_nodes": sum(cluster["nodes"] for cluster in mock_hpc_data.values()),
        "avg_cpu_utilization": sum(cluster["cpu_utilization"] for cluster in mock_hpc_data.values()) / len(mock_hpc_data)
    }

@router.post("/submit-job", response_model=QuantumDataResponse)
async def submit_quantum_job(request: QuantumDataRequest):
    """
    Submit a new quantum computing job
    """
    job_id = f"qjob_{datetime.now().strftime('%Y%m%d_%H%M%S')}"
    
    # Simulate job processing
    await asyncio.sleep(0.1)  # Simulate processing time
    
    return QuantumDataResponse(
        job_id=job_id,
        status="submitted",
        quantum_data={
            "provider": request.quantum_provider,
            "qubits_requested": 10,
            "circuit_depth": 50
        },
        hpc_metrics={
            "cluster": request.hpc_cluster,
            "nodes_allocated": 5,
            "estimated_runtime": "15 minutes"
        },
        timestamp=datetime.now(),
        estimated_completion=datetime.now()
    )

@router.get("/job-status/{job_id}", response_model=JobStatusResponse)
async def get_job_status(job_id: str):
    """
    Get status of a specific quantum job
    """
    # Simulate different job states
    import random
    statuses = ["submitted", "running", "completed", "failed"]
    status = random.choice(statuses)
    
    return JobStatusResponse(
        job_id=job_id,
        status=status,
        progress=random.uniform(0, 100),
        quantum_qubits=random.randint(5, 20),
        hpc_utilization=random.uniform(20, 90),
        error_message=None if status != "failed" else "Quantum decoherence detected"
    )

@router.get("/system-health", response_model=SystemHealthResponse)
async def get_system_health():
    """
    Get overall system health status
    """
    return SystemHealthResponse(
        quantum_systems={name: data["status"] for name, data in mock_quantum_data.items()},
        hpc_clusters={name: data["status"] for name, data in mock_hpc_data.items()},
        api_status="healthy",
        last_updated=datetime.now()
    )

@router.get("/monitoring/metrics")
async def get_monitoring_metrics():
    """
    Get real-time monitoring metrics
    """
    return {
        "quantum_metrics": {
            "total_jobs": 15,
            "success_rate": 94.2,
            "avg_execution_time": "12.5 minutes",
            "queue_wait_time": "3.2 minutes"
        },
        "hpc_metrics": {
            "total_jobs": 45,
            "success_rate": 98.7,
            "avg_execution_time": "8.3 minutes",
            "resource_utilization": 72.1
        },
        "system_metrics": {
            "api_response_time": "45ms",
            "database_connections": 12,
            "memory_usage": "2.1GB",
            "cpu_usage": "35%"
        },
        "timestamp": datetime.now()
    }

@router.get("/remediation/status")
async def get_remediation_status():
    """
    Get self-healing and remediation system status
    """
    return {
        "self_healing": {
            "enabled": True,
            "last_check": datetime.now(),
            "issues_detected": 2,
            "issues_resolved": 1,
            "success_rate": 95.5
        },
        "remediation_actions": [
            {
                "action_id": "rmd_001",
                "type": "quantum_error_correction",
                "status": "completed",
                "timestamp": datetime.now()
            },
            {
                "action_id": "rmd_002", 
                "type": "hpc_resource_optimization",
                "status": "in_progress",
                "timestamp": datetime.now()
            }
        ],
        "recommendations": [
            "Consider increasing quantum error correction for job qjob_20240908_143022",
            "HPC cluster aws_hpc shows high memory utilization - consider scaling"
        ]
    }
